driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/ClientSessionBinding.java

```diff
@@ -107,28 +84,49 @@ public OperationContext getOperationContext() {
         return wrapped.getOperationContext();
     }
 
-    private void getPinnedConnectionSource(final boolean isRead, final SingleResultCallback<AsyncConnectionSource> callback) {
+    @Override
+    public void getReadConnectionSource(final SingleResultCallback<AsyncConnectionSource> callback) {
+        getConnectionSource(wrapped::getReadConnectionSource, callback);
+    }
+
+    @Override
+    public void getReadConnectionSource(final int minWireVersion, final ReadPreference fallbackReadPreference,
+            final SingleResultCallback<AsyncConnectionSource> callback) {
+        getConnectionSource(wrappedConnectionSourceCallback ->
+                wrapped.getReadConnectionSource(minWireVersion, fallbackReadPreference, wrappedConnectionSourceCallback),
+                callback);
+    }
+
+    public void getWriteConnectionSource(final SingleResultCallback<AsyncConnectionSource> callback) {
+        getConnectionSource(wrapped::getWriteConnectionSource, callback);
+    }
+
+    private interface AsyncConnectionSourceSupplier {
+        void getConnectionSource(SingleResultCallback<AsyncConnectionSource> callback);
+    }
+
+    private void getConnectionSource(final AsyncConnectionSourceSupplier connectionSourceSupplier,
```

Comment:
This is easier to understand in the sync version, but the algorithm is the same.